cardano-node/src/Cardano/Node/Run.hs

```diff
@@ -219,8 +229,29 @@ handleSimpleNode p trace nodeTracers npm onKernel = do
     (isProducer nc)
     (customiseChainDbArgs $ dbValidation npm)
     identity -- No NodeParams customisation
-    $ \_registry nodeKernel -> onKernel nodeKernel
+    $ \registry nodeKernel -> do
+      case (stopOnSlotSynced npm, sfds) of
+        (MaxSlotNo maxSlot, InternalShutdown _r shutFd) -> do
+          traceWith (trTransformer TextualRepresentation MaximalVerbosity $
+                      severityNotice trace)
+            ("will terminate upon reaching " <> (pack $ show maxSlot) :: Text)
+          spawnSlotLimitTerminator registry (Node.getChainDB nodeKernel) maxSlot shutFd
+        _ -> pure ()
+      onKernel nodeKernel
  where
+  spawnSlotLimitTerminator
+    :: ResourceRegistry IO -> ChainDB.ChainDB IO blk -> SlotNo -> Fd -> IO ()
+  spawnSlotLimitTerminator registry chainDB maxSlot (Fd shutFd) =
+    void $ onEachChange registry identity Nothing
+      (getPoint <$> ChainDB.getTipPoint chainDB) $
```

Comment:
Use `pointSlot` here so you can pattern match on `WithOrigin SlotNo`

Reply:
Done.